ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceFileManager.swift

```diff
@@ -119,22 +119,28 @@ public class ResourceFileManager {
   internal func copyWithOverwrite(from source: URL, to destination: URL) throws {
     let fileManager = FileManager.default
 
-    // For now, we'll always allow overwriting.
-    if fileManager.fileExists(atPath: destination.path) {
-      try fileManager.removeItem(at: destination)
-    }
-
     // If we've been provided a security-scoped resource URL,
     // it needs special handling.  This function needs to accept
     // both scoped & non-scoped URLs.
-    if source.startAccessingSecurityScopedResource() { // only succeeds if scoped
-      // The Swift version of 'finally'.
-      defer { source.stopAccessingSecurityScopedResource() }
-      try fileManager.copyItem(at: source, to: destination)
-    } else {
-      // Not scoped?  No problem!
-      try fileManager.copyItem(at: source, to: destination)
+    let isSecurityScoped = source.startAccessingSecurityScopedResource()
+    defer {
+      if isSecurityScoped {
+        source.stopAccessingSecurityScopedResource()
+      }
     }
+
+    // Build a temporary copy of the file that will replace the destination file
+    // so that we can do the overwrite atomically, ensuring that we can't end
+    // in a state where the destination was erased but never got re-written due
+    // to a copy error.
+    let tempFile = fileManager.temporaryDirectory.appendingPathComponent("temp.file")
+    try? fileManager.removeItem(at: tempFile) // Just in case the replace fails.
```

Comment:
Is this remove in case a previous attempt failed?

Reply:
> ```
> // Just in case the replace fails.
> ```

Yes.  `replaceItem`, further down, removes the prior attempt's source file, so it won't be present after successful copies.

Perhaps more accurately in regard to `replaceItem`, it replaces the source file's filename... and the contents of whatever file may have previously held the new filename.

Reply:
Ok, so I think the comment is confusing me. It says "Just in case the replace fails." But it is actually a cleanup step in case a previous attempt failed -- before attempting to copy again?